### ISSUE TITLE
Question about Training Process and Model Checkpoints

### ISSUE THREAD

**Issue by YHX021014** [kuleshov-group/caduceus#71]
Hi,

I have a few questions regarding the training process based on my script:

1. During training, I'm observing that val/loss=nan.0, val/perplexity=nan.0, test/loss=nan.0, and test/perplexity=nan.0, even if I got a continues declining loss. I believe this might be a serious issue. Have you encountered something similar before? Any suggestions on how to resolve this?

Here are my traing script and log files:

traing script:
```
cd ../ || exit  # Go to the root directory of the repo

export HYDRA_FULL_ERROR=1

NUM_DEVICES=4

# Run script
SEQLEN=32768
MAX_STEPS=10000
D_MODEL=256
N_LAYER=8
LR="8e-3"
BIDIRECTIONAL_STRATEGY="add"
BIDIRECTIONAL_WEIGHT_TIE="true"
RCPS="true"
RC_AUG="false"

BATCH_SIZE=$(( 1048576 / SEQLEN ))
SEQLEN_DIS="$(echo "scale=0; ${SEQLEN} / 1000" | bc)k"
WANDB_NAME="caduceus-ps_seqlen-${SEQLEN_DIS}_d_model-${D_MODEL}_n_layer-${N_LAYER}_lr-${LR}"
HYDRA_RUN_DIR="./outputs/pretrain/hg38/${WANDB_NAME}"

mkdir -p "${HYDRA_RUN_DIR}"
python -m train \
  experiment=hg38/hg38 \
  callbacks.model_checkpoint_every_n_steps.every_n_train_steps=500 \
  dataset.max_length=${SEQLEN} \
  dataset.batch_size=$(( BATCH_SIZE / NUM_DEVICES / 8 )) \
  dataset.mlm=true \
  dataset.mlm_probability=0.15 \
  dataset.rc_aug="${RC_AUG}" \
  model="caduceus" \
  model.config.d_model=${D_MODEL} \
  model.config.n_layer=${N_LAYER} \
  model.config.bidirectional=true \
  model.config.bidirectional_strategy=${BIDIRECTIONAL_STRATEGY} \
  model.config.bidirectional_weight_tie=${BIDIRECTIONAL_WEIGHT_TIE} \
  model.config.rcps=${RCPS} \
  optimizer.lr="${LR}" \
  train.global_batch_size=${BATCH_SIZE} \
  trainer.max_steps=${MAX_STEPS} \
  trainer.devices=${NUM_DEVICES} \
  +trainer.val_check_interval=$(( MAX_STEPS / 10 )) \
  wandb.group=pretrain_hg38 \
  wandb.name="${WANDB_NAME}" \
  wandb.job_type="caduceus" \
  hydra.run.dir="${HYDRA_RUN_DIR}"
```
log file:
https://drive.google.com/file/d/100REdlCnIYiQGgVcD_OwdCAj_Ku0yM9V/view?usp=sharing

![Image](https://github.com/user-attachments/assets/59dc37fa-9de1-489c-bf9a-b9b2b4b0262e)

![Image](https://github.com/user-attachments/assets/121746b8-969d-4fae-9320-a1fe2b4e0bfc)
```
Will assume defaults for version 1.1
  @hydra.main(config_path="configs", config_name="config.yaml")
CONFIG
├── train
│   └── seed: 2222                                                                                                                                                                
│       interval: step                                                                                                                                                            
│       monitor: test/loss                                                                                                                                                        
│       mode: min                                                                                                                                                                 
│       ema: 0.0                                                                                                                                                                  
│       test: true                                                                                                                                                                
│       debug: false                                                                                                                                                              
│       ignore_warnings: false                                                                                                                                                    
│       optimizer_param_grouping:                                                                                                                                                 
│         bias_weight_decay: false                                                                                                                                                
│         normalization_weight_decay: false                                                                                                                                       
│       state:                                                                                                                                                                    
│         mode: null                                                                                                                                                              
│         n_context: 0                                                                                                                                                            
│         n_context_eval: 0                                                                                                                                                       
│       ckpt: checkpoints/last.ckpt                                                                                                                                               
│       disable_dataset: false                                                                                                                                                    
│       validate_at_start: false                                                                                                                                                  
│       pretrained_model_path: null                                                                                                                                               
│       pretrained_model_strict_load: true                                                                                                                                        
│       pretrained_model_state_hook:                                                                                                                                              
│         _name_: null                                                                                                                                                            
│       post_init_hook:                                                                                                                                                           
│         _name_: null                                                                                                                                                            
│       layer_decay:                                                                                                                                                              
│         _name_: null                                                                                                                                                            
│         decay: 0.7                                                                                                                                                              
│       gpu_mem: 25                                                                                                                                                               
│       global_batch_size: 32                                                                                                                                                     
│                                                                                                                                                                                 
├── wandb
│   └── project: dna                                                                                                                                                              
│       group: pretrain_hg38                                                                                                                                                      
│       job_type: caduceus                                                                                                                                                        
│       mode: online                                                                                                                                                              
│       name: caduceus-ps_seqlen-32k_d_model-256_n_layer-8_lr-8e-3                                                                                                                
│       save_dir: .                                                                                                                                                               
│       id: caduceus-ps_seqlen-32k_d_model-256_n_layer-8_lr-8e-3                                                                                                                  
│                                                                                                                                                                                 
├── trainer
│   └── _target_: pytorch_lightning.Trainer                                                                                                                                       
│       devices: 4                                                                                                                                                                
│       accelerator: gpu                                                                                                                                                          
│       accumulate_grad_batches: 8                                                                                                                                                
│       max_epochs: null                                                                                                                                                          
│       gradient_clip_val: 1.0                                                                                                                                                    
│       log_every_n_steps: 10                                                                                                                                                     
│       limit_train_batches: 1.0                                                                                                                                                  
│       limit_val_batches: 0.125                                                                                                                                                  
│       num_sanity_val_steps: 2                                                                                                                                                   
│       num_nodes: 1                                                                                                                                                              
│       max_steps: 10000                                                                                                                                                          
│       precision: 16                                                                                                                                                             
│       val_check_interval: 1000                                                                                                                                                  
│                                                                                                                                                                                 
├── dataset
│   └── _name_: hg38                                                                                                                                                              
│       bed_file: null                                                                                                                                                            
│       fasta_file: null                                                                                                                                                          
│       dataset_name: hg38                                                                                                                                                        
│       tokenizer_name: char                                                                                                                                                      
│       cache_dir: null                                                                                                                                                           
│       max_length: 32768                                                                                                                                                         
│       add_eos: true                                                                                                                                                             
│       batch_size: 1                                                                                                                                                             
│       batch_size_eval: 2                                                                                                                                                        
│       num_workers: 12                                                                                                                                                           
│       shuffle: true                                                                                                                                                             
│       max_length_val: 32768                                                                                                                                                     
│       max_length_test: 32768                                                                                                                                                    
│       pad_max_length: null                                                                                                                                                      
│       rc_aug: false                                                                                                                                                             
│       use_fixed_len_val: false                                                                                                                                                  
│       mlm: true                                                                                                                                                                 
│       mlm_probability: 0.15                                                                                                                                                     
│                                                                                                                                                                                 
├── optimizer
│   └── _name_: adamw                                                                                                                                                             
│       lr: 0.008                                                                                                                                                                 
│       weight_decay: 0.1                                                                                                                                                         
│       betas:                                                                                                                                                                    
│       - 0.9                                                                                                                                                                     
│       - 0.95                                                                                                                                                                    
│                                                                                                                                                                                 
├── scheduler
│   └── _name_: cosine_warmup_timm                                                                                                                                                
│       t_in_epochs: false                                                                                                                                                        
│       t_initial: 9000.0                                                                                                                                                         
│       lr_min: 0.0001                                                                                                                                                            
│       warmup_lr_init: 1.0e-06                                                                                                                                                   
│       warmup_t: 1000.0                                                                                                                                                          
│       warmup_prefix: true                                                                                                                                                       
│                                                                                                                                                                                 
├── callbacks
│   └── learning_rate_monitor:                                                                                                                                                    
│         logging_interval: step                                                                                                                                                  
│       timer:                                                                                                                                                                    
│         step: true                                                                                                                                                              
│         inter_step: false                                                                                                                                                       
│         epoch: true                                                                                                                                                             
│         val: true                                                                                                                                                               
│       params:                                                                                                                                                                   
│         total: true                                                                                                                                                             
│         trainable: true                                                                                                                                                         
│         fixed: true                                                                                                                                                             
│       model_checkpoint:                                                                                                                                                         
│         monitor: test/loss                                                                                                                                                      
│         mode: min                                                                                                                                                               
│         save_top_k: 1                                                                                                                                                           
│         save_last: false                                                                                                                                                        
│         dirpath: checkpoints/                                                                                                                                                   
│         filename: test/loss                                                                                                                                                     
│         auto_insert_metric_name: false                                                                                                                                          
│         verbose: true                                                                                                                                                           
│       model_checkpoint_every_n_steps:                                                                                                                                           
│         monitor: train/loss                                                                                                                                                     
│         mode: min                                                                                                                                                               
│         save_top_k: 0                                                                                                                                                           
│         save_last: true                                                                                                                                                         
│         dirpath: checkpoints/                                                                                                                                                   
│         filename: train/loss                                                                                                                                                    
│         auto_insert_metric_name: false                                                                                                                                          
│         verbose: true                                                                                                                                                           
│         every_n_train_steps: 500                                                                                                                                                
│                                                                                                                                                                                 
├── task
│   └── _name_: lm                                                                                                                                                                
│       loss:                                                                                                                                                                     
│         _name_: cross_entropy                                                                                                                                                   
│         ignore_index: 4                                                                                                                                                         
│       torchmetrics:                                                                                                                                                             
│       - perplexity                                                                                                                                                              
│       - num_tokens                                                                                                                                                              
│                                                                                                                                                                                 
├── encoder
│   └── None                                                                                                                                                                      
├── decoder
│   └── None                                                                                                                                                                      
├── loader
│   └── num_workers: 96                                                                                                                                                           
│       pin_memory: true                                                                                                                                                          
│       drop_last: true                                                                                                                                                           
│                                                                                                                                                                                 
└── model
    └── _name_: caduceus_lm                                                                                                                                                       
        config:                                                                                                                                                                   
          _target_: caduceus.configuration_caduceus.CaduceusConfig                                                                                                                
          d_model: 256                                                                                                                                                            
          n_layer: 8                                                                                                                                                              
          vocab_size: 12                                                                                                                                                          
          ssm_cfg:                                                                                                                                                                
            d_state: 16                                                                                                                                                           
            d_conv: 4                                                                                                                                                             
            expand: 2                                                                                                                                                             
            dt_rank: auto                                                                                                                                                         
            dt_min: 0.001                                                                                                                                                         
            dt_max: 0.1                                                                                                                                                           
            dt_init: random                                                                                                                                                       
            dt_scale: 1.0                                                                                                                                                         
            dt_init_floor: 0.0001                                                                                                                                                 
            conv_bias: true                                                                                                                                                       
            bias: false                                                                                                                                                           
            use_fast_path: true                                                                                                                                                   
          rms_norm: true                                                                                                                                                          
          fused_add_norm: true                                                                                                                                                    
          residual_in_fp32: false                                                                                                                                                 
          pad_vocab_size_multiple: 8                                                                                                                                              
          norm_epsilon: 1.0e-05                                                                                                                                                   
          initializer_cfg:                                                                                                                                                        
            initializer_range: 0.02                                                                                                                                               
            rescale_prenorm_residual: true                                                                                                                                        
            n_residuals_per_layer: 1                                                                                                                                              
          bidirectional: true                                                                                                                                                     
          bidirectional_strategy: add                                                                                                                                             
          bidirectional_weight_tie: true                                                                                                                                          
          rcps: true                                                                                                                                                              
          complement_map: null                                                                                                                                                    
                                                                                                                                                                                  
Global seed set to 2222
```

2. The loss curve during training looks quite strange. Could this be an indication of a problem?

I got you some screenshots of my pretraining process
![Image](https://github.com/user-attachments/assets/0e3fdd9c-19d9-4ea0-8ee1-21b494cd84eb)

![Image](https://github.com/user-attachments/assets/34be2a07-1727-4553-81f8-cc826f79badc)

![Image](https://github.com/user-attachments/assets/9f04a52c-d506-477a-a578-60607fa7ff78)

3. Could you clarify the difference between the saved models `checkpoints/last.ckpt` and `checkpoints/test/loss.ckpt`? What does each checkpoint represent in terms of the model state? Also, since my test/loss has been consistently nan, does this mean that the `checkpoints/test/loss.ckpt` is essentially unusable? Furthermore, due to the instability of my training, the model saved in `checkpoints/last.ckpt` may not necessarily be the best performing one.

I would greatly appreciate it if you could offer some insights and suggestions regarding these issues. Your expertise and guidance would be extremely helpful in resolving these problems and ensuring the success of my training process. Thank you so much for your time and consideration.

Best regards！

**Comment by nine-ce** [comment]
I also encountered the same problem. May I ask if you have solved it now? Is NaN normal?

**Comment by leannmlindsey** [comment]
I am not one of the authors, just a user, but my experience has been if your training loss is Nan, it is not normal and the easiest way to fix it is to lower your learning rate.

**Comment by yair-schiff** [comment]
@YHX021014 and @nine-ce , apologies for the delayed response. I agree with @leannmlindsey , this is usually a sign that you hit NaN somewhere during training and LR is too high.

were your able to work around this issue with lower learning rates?

**Comment by yair-schiff** [comment]
For the most part, your script looks correct. The only differences I can see is that i used LR `8e-3` and for this line:
> `BATCH_SIZE=$(( 524288 / SEQLEN ))`

I used `1048576` instead of `524288`. I did not hit NaN's when running my pre-training here
